File: src/containers/Experience.js
import _ from "lodash";
import ExperienceRow from "../components/ExperienceRow";
import SkeletonPlaceholder from "../components/shared/SkeletonPlaceholder";
import config from "../config";
import experienceData from "../data/experiences.json";
import { useCachedFetch } from "../utils/cachedFetchHook";
import { sortByDateRange } from "../utils/sort";

function Experience() {
  const PLACEHOLDERS = Array(2).fill(0);

  let { loading, data, error } = useCachedFetch(
    `${config.API_ENDPOINT}/experiences`,
    config.CACHE_TOGGLE
  );

  if (error) {
    console.log(error);
    console.log("Loading static data as fallback...");
    data = experienceData;
  }

  return (
    <>
      {loading || config.PRE_RENDERING ? (
        PLACEHOLDERS.map((_, index) => (
          <div key={index} className="mb-4">
            <SkeletonPlaceholder rows={6} ready={!loading && !config.PRE_RENDERING} />
          </div>
        ))
      ) : !_.isEmpty(data) ? (
        sortByDateRange(data).map((d, index) => (
          <ExperienceRow
            key={d.id}
            title={d.jobTitle}
            company={d.company}
            desc={d.desc}
            url={d.url}
            location={d.location}
            date={d.date}
            lastRow={index + 1 === _.size(data)}
          />
        ))
      ) : (
        <></>
      )}
    </>
  );
}

export default Experience;

File: db/models/Experience.go
package models

import "time"

type Experience struct {
	Id               string `bson:"_id,omitempty" json:"id,omitempty"`
	JobTitle         string `bson:"jobTitle,omitempty" json:"jobTitle"`
	Company          string `bson:"company,omitempty" json:"company"`
	Desc             string `bson:"desc,omitempty" json:"desc"`
	Url              string `bson:"url,omitempty" json:"url"`
	Location         string `bson:"location,omitempty" json:"location"`
	Date             string `bson:"date,omitempty" json:"date"`
	CreatedTimestamp string `bson:"createdTimestamp,omitempty" json:"createdTimestamp"`
}

// Constructor
func NewExperience() Experience {
	instance := Experience{}
	instance.JobTitle = ""
	instance.Company = ""
	instance.Desc = ""
	instance.Url = ""
	instance.Location = ""
	instance.Date = time.Now().Format("2006")
	instance.CreatedTimestamp = time.Now().Format("2006-01-02T15:04:05Z07:00")
	return instance
}
